### ISSUE TITLE
Built-in Symbol.iterator methods return IterableIterator<T> extending Iterator<T>, not Iterator<T, void>

### ISSUE THREAD

**Issue by lionel-rowe** [microsoft/TypeScript#52998]
# lib Update Request

The built-in `Symbol.iterator` methods return `IterableIterator<T>` extending `Iterator<T>`, not the expected `Iterator<T, void>`. For example:

```ts
const char = string[Symbol.iterator]().next().value
```

The expected type of `char` is `string | undefined`, but instead it's `any`. This isn't correct — it will never be anything other than a string (strictly a single Unicode char) or `undefined`.

This looks to be related to https://github.com/microsoft/TypeScript/issues/33353, which was closed due to backwards-compatibility concerns; however, it could presumably fixed without changing the default types of `IterableIterator`, instead changing its definition to take an extra type parameter:

```diff
- interface IterableIterator<T> extends Iterator<T> {
+ interface IterableIterator<T, TReturn = any> extends Iterator<T, TReturn> {
      [Symbol.iterator](): IterableIterator<T>;
  }
```

And then changing all the built-ins (`String`, `Array`, etc) like so:

```diff
  interface String {
      /** Iterator */
-     [Symbol.iterator](): IterableIterator<string>;
+     [Symbol.iterator](): IterableIterator<string, void>;
  }
```

Could that work as a fix and be sufficiently backward-compatible?

## Configuration Check

My compilation *target* is `ES2020` and my *lib* is `the default`.

## Missing / Incorrect Definition

Various — any built-in JS type with a `Symbol.iterator` method, such as `Array`, `String`, `Map`, `Set`, `Uint8Array`, etc.

## Sample Code

```ts
const char = string[Symbol.iterator]().next().value
// expected type: string | undefined
// actual type: any

const definitelyChar = nonEmptyString[Symbol.iterator]().next().value!
// expected type: string
// actual type: any
```

## Documentation Link

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/@@iterator

**Comment by fatcerberus** [comment]
Somewhat of a tangent but `void` is not a synonym for `undefined` and shouldn’t be treated as such.  See https://github.com/microsoft/TypeScript/issues/42709.

**Comment by lionel-rowe** [comment]
> Somewhat of a tangent but `void` is not a synonym for `undefined` and shouldn’t be treated as such. See #42709.

Thanks for the correction! I've fixed my original comment to be more accurate about this.

**Comment by fatcerberus** [comment]
> The expected type of char is `string | void`

Again, I don’t think you really want this.  You can’t narrow `void` out of a union.  If your argument is that the value can be `undefined` then you should use `undefined` explicitly.

**Comment by lionel-rowe** [comment]
> If your argument is that the value can be `undefined` then you should use `undefined` explicitly.

Semantically, `string | void` makes more sense to me than  `string | undefined`, by analogy with generator functions that don't explicitly return something:

```ts
// type of iterateStr is Generator<string, void, undefined>, i.e. TReturn is void
function* iterateStr(str: string) {
    yield* str
}
```

> You can’t narrow `void` out of a union.

Seems to work fine to me? [Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEEQBd4A8Bc8DOyYEsA7Ac3gB94A3Ae32HgCgH8AzeACmQE8AHEatmngBeUfADkuAiXEBKeAG8G8FengB6dTjxFiDAL5MkqLiPQBCDVqm6Gx+AC8sAUTSQArqAA83PgPQANFS0wAB8VtrSekA)

```ts
declare let x: string | void 

if (typeof x === 'string') {
    x // string
}

let y = x! // string
let z: Exclude<typeof x, void> // string
```

I'm not married to using `void` over `undefined` if the latter will prevent some class of bugs, though.

**Comment by fatcerberus** [comment]
The problem with `string | void`, is that `void` is an implicit promise that the code receiving a value of that type *won't access it at all*.  More critically, `() => T` is assignable to `() => void` for all `T`, which means the value is not even guaranteed to be `undefined`.  `string | void` is thus conceptually unsound because you have to access the value to narrow it.  It's more-or-less an `any` that you're not allowed to examine.

More concretely, here's the issue with `void`:
```ts
const foo: () => void = () => 777;
const bar: () => number = () => 1206;
const x: number | void = Math.random() > 0.5 ? foo() : bar;
// note: by calling foo which has a void return type, we've promised to discard its return value.
// but now there's a problem: we have no way to know whether x is garbage or not!
```

In other words, in the context of larger union type it's basically just an `any` with extra steps.

**Comment by lionel-rowe** [comment]
> More concretely, here's the issue with void:

Wow, `void` is a lot weirder than I thought... I've reworded the OP to use `undefined` exclusively.

Is the fact that union types like `string | void` can be checked for truthiness considered a bug, then? It certainly seems like one:

```ts
const fn: () => void = () => 1
let x: string | void = 2 > 1 ? fn() : 'str'

if (x) {
    x.trim() // passes type checking but throws runtime error
}
```

 If this is considered a bug, is there a live issue for it where we can split of this discussion?

**Comment by fatcerberus** [comment]
I linked it above: https://github.com/microsoft/TypeScript/issues/42709

**Comment by MuTsunTsai** [comment]
The same issue happens on `Set.prototype.values()` as well:
```ts
const set = new Set<string>();
const v = set.values().next().value;
// should be string | undefined, but get any
```
[Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBBCmsC8MzwO4wMqIDzQCcBLMAcwD4AKASgG4BYAKFElgDcYUEoA6NgQwA2AV3gQaPNAA8oEgSPgNGAemVwAFiGGCAJjABG8OFGJkYAHxjCwO+ADMS8HQBoDw2KUQx+YAJ5A)

**Comment by lionel-rowe** [comment]
Draft PR up here: https://github.com/microsoft/TypeScript/pull/56502

Having thought some more about it though, maybe it'd be better to have an entirely new interface that doesn't extend `Iterator`. In addition to defaulting `TReturn` to `any`,  `Iterator` also has two methods that are never present on the built-in iterable iterators:

```ts
    return?(value?: TReturn): IteratorResult<T, TReturn>;
    throw?(e?: any): IteratorResult<T, TReturn>;
```

That widens the discrepancy between compile-time type checking and runtime type guarantees even further:

```ts
const arr = [1]
const iter = arr.values()

const next = iter.next()
if (!next.done) next.value  // ✅ type = number
next.value // ❌ type = any (should be number | undefined)

iter.nonExistentMethod?.() // ✅ Property 'nonExistentMethod' does not exist on type 'IterableIterator<number>'.
iter.return?.() // ❌ TS allows this, even though the `return` method will never exist (absent monkey-patching etc)
iter.throw?.() // ❌ TS allows this, even though the `throw` method will never exist (absent monkey-patching etc)
```

Instead, we could have something like this:

```diff
interface Iterator<T, TReturn = any, TNext = undefined> {
    next(...args: [] | [TNext]): IteratorResult<T, TReturn>;
    return?(value?: TReturn): IteratorResult<T, TReturn>;
    throw?(e?: any): IteratorResult<T, TReturn>;
}

interface IterableIterator<T> extends Iterator<T> {
    [Symbol.iterator](): IterableIterator<T>;
}

+ interface BuiltInIterableIterator<T> {
+     next(): IteratorResult<T, undefined>;
+     [Symbol.iterator](): BuiltInIterableIterator<T>;
+ }

interface Array<T> {
-     [Symbol.iterator](): IterableIterator<T>;
+     [Symbol.iterator](): BuiltInIterableIterator<T>;
-     entries(): IterableIterator<[number, T]>;
+     entries(): BuiltInIterableIterator<[number, T]>;
-     keys(): IterableIterator<number>;
+     keys(): BuiltInIterableIterator<number>;
-     values(): IterableIterator<T>;
+     values(): BuiltInIterableIterator<T>;
}
```

Not sure if `BuiltInIterableIterator` is the best name though — maybe `GenericIterableIterator` (not to be confused with TS generics) or something else?

**Comment by fatcerberus** [comment]
Not sure I agree on the second point re: removing the nonexistent methods simply because they never exist.  Code may want to deal with iterators generically and not have to care whether it’s a built-in iterator or not.  `iter.return?.()` should still be allowed, IMO.

Future versions of ECMAScript may also decide to add those methods, too, so people might be writing `iter.return?.()` *today*, just for the sake of future-proofing.

**Comment by lionel-rowe** [comment]
> Not sure I agree on the second point re: removing the nonexistent methods simply because they never exist. Code may want to deal with iterators generically and not have to care whether it’s a built-in iterator or not. `iter.return?.()` should still be allowed, IMO.

I'm not suggesting function parameter types in library code should be changed, only the types of the built-in iterables themselves. Userland code, on the other hand, can be as strict or lenient as it likes.

```ts
interface Iterator<T, TReturn = any, TNext = undefined> {
    next(...args: [] | [TNext]): IteratorResult<T, TReturn>;
    return?(value?: TReturn): IteratorResult<T, TReturn>;
    throw?(e?: any): IteratorResult<T, TReturn>;
}

interface IterableIterator<T> extends Iterator<T> {
    [Symbol.iterator](): IterableIterator<T>;
}

interface BuiltInIterableIterator<T> {
    next(): IteratorResult<T, undefined>;
    [Symbol.iterator](): BuiltInIterableIterator<T>;
}

// an existing library function that's always accepted `IterableIterator<number>`; signature remains unchanged
function someLibraryFn(iter: IterableIterator<number>) {
    iter.throw?.()
    iter.return?.()

    return iter.next().value
}

declare const iter: BuiltInIterableIterator<number>

// works with no issues
// `BuiltInIterableIterator<number>` is still assignable to IterableIterator<number>
// type of `val` remains `any`
const val = someLibraryFn(iter)
```

Anyhow, my PR https://github.com/microsoft/TypeScript/pull/56517 is failing CI due to `Type 'void' is not assignable to type 'undefined'.` Per the discussion earlier in this issue thread, I don't think typing `next`'s return type as `IteratorResult<T, void>;` is an appropriate solution; in which case, it would seem like this fix is blocked pending https://github.com/microsoft/TypeScript/issues/36239.

**Comment by lionel-rowe** [comment]
As this may not get fixed any time soon, I'll leave this here too: my main use case for this is wanting an easy way to grab the first element from an iterable (often the first UTF-8 char of a string). The most ergonomic way is `[...str][0]`, but iterating the whole string just to get the first item seems pretty wasteful. `str[Symbol.iterator]().next().value` is only a shade more verbose, but loses all type information in the process due to this issue.

What I've settled on since I opened this issue is the following:

```ts
let char = ''
for (const ch of str) {
    char = ch
    break
}
```

It looks a little funky, but it's reasonably concise, performant (doesn't iterate the whole string), and also has the advantage of forcing you to choose a default (in this case `''`) in case the string is empty.

An alternative using an IIFE, which is even funkier, but has the advantage of allowing you to declare the variable with `const`:

```ts
const char = (() => {
    for (const ch of str) return ch
    return ''
})()
```
